### PR TITLE
protoc extensions supported.

### PR DESCRIPTION
**Pull request by noel-yap**
Custom protoc options (https://developers.google.com/protocol-buffers/docs/proto3#custom_options) are implemented via extensions (https://developers.google.com/protocol-buffers/docs/reference/java-generated#extension). If a custom option is to be processed by a protoc plugin, the plugin code must register the extension. For example (from the docs):

```
ExtensionRegistry registry = ExtensionRegistry.newInstance();
registry.add(Baz.fooExt);
Foo foo = Foo.parseFrom(input, registry);
```

### CODE REVIEW

**Comment by rmichela** [comment]
Thanks for the PR! Could you elaborate on what this does?

File: jprotoc/src/main/java/com/salesforce/jprotoc/ProtocPlugin.java [hunk 1]
```diff
@@ -40,13 +41,25 @@ public static void generate(@Nonnull Generator generator) {
      * @param generators The list of generators to run.
      */
     public static void generate(@Nonnull List<Generator> generators) {
+        generate(generators, Collections.emptyList());
+    }
+
+    public static void generate(
```
Comment (rmichela) [inline jprotoc/src/main/java/com/salesforce/jprotoc/ProtocPlugin.java]: Needs JavaDoc.

File: jprotoc/src/main/java/com/salesforce/jprotoc/ProtocPlugin.java [hunk 2]
```diff
@@ -40,13 +41,25 @@ public static void generate(@Nonnull Generator generator) {
...
+            @Nonnull List<Generator> generators, List<GeneratedExtension> extensions) {
         Preconditions.checkNotNull(generators, "generators");
         Preconditions.checkArgument(!generators.isEmpty(), "generators.isEmpty()");
+        Preconditions.checkNotNull(extensions, "extensions");
+
+        ExtensionRegistry extensionRegistry = ExtensionRegistry.newInstance();
```
Comment (rmichela) [inline jprotoc/src/main/java/com/salesforce/jprotoc/ProtocPlugin.java]: Please add a comment explaining what the `extensionRegistry` does.

File: jprotoc/src/main/java/com/salesforce/jprotoc/ProtocPlugin.java (hunk 1 above)
Comment (noel-yap) [reply jprotoc/src/main/java/com/salesforce/jprotoc/ProtocPlugin.java]: Will do.

Comment (noel-yap) [reply jprotoc/src/main/java/com/salesforce/jprotoc/ProtocPlugin.java]: Done.

File: jprotoc/src/main/java/com/salesforce/jprotoc/ProtocPlugin.java (hunk 2 above)
Comment (noel-yap) [reply jprotoc/src/main/java/com/salesforce/jprotoc/ProtocPlugin.java]: Done.

File: jprotoc/README.md [hunk 3]
```diff
@@ -130,4 +130,43 @@ Using the Jdk8 Protoc generator
...
+    
+    extend google.protobuf.FileOptions {
+        ServerOptions server = 50621;
+    }
+    
+    message ServerOptions {
+        // Java classname
+        string name = 1;
+    }
+    ```
```
Comment (rmichela) [inline jprotoc/README.md]: Do I have to run the protobuf generator as part of the build process for my protobuf plugin? Do I need to add this to my plugin project's POM?

File: jprotoc/README.md [hunk 4]
```diff
@@ -130,4 +130,43 @@ Using the Jdk8 Protoc generator
...
+
```
Comment (rmichela) [inline jprotoc/README.md]: Do I have to run the protobuf generator as part of the build process for my protobuf plugin? Do I need to add this to my plugin project's POM?

Comment (noel-yap) [reply jprotoc/README.md]: I'm not familiar with Maven but I don't see any reason anything in the build would have to change. All the processing is done at `protoc` run-time when `protoc` calls the `protoc` plugin.

**Comment by noel-yap** [comment]
I'm not understanding the output of the code coverage report. Some lines (eg `ProtocPlugin.java:44`) are marked red but I don't see how that could be since the entrypoint should have been covered by some previous test. Were the reports added later and there's actually no test covering `ProtocPlugin.generate()`? I'm not finding any test code that calls it.

**Comment by rmichela** [comment]
Ignore the CodeCov report. I'm still trying to get it working right. 😉

File: jprotoc/README.md (hunk 4 above)
Comment (rmichela) [reply jprotoc/README.md]: Maybe I'm not understanding correctly. Where does the type `com.example.proto.options.ServerOptionsProto.server` come from in the example below? Perhaps you could add an elaborated example in `jprotoc-test`?

Comment (noel-yap) [reply jprotoc/README.md]: Oh, that's generated by `protoc`. Those not creating custom options don't need to run `protoc` to generate the related Java code for the option. I'm assuming those creating custom options would know how to run `protoc` to do that. I can add a blurb about running `protoc` if you want.

Comment (rmichela) [reply jprotoc/README.md]: I see...My JProtoc Protobuf generator project would itself contain Protobuf generated code (it's proto all the way down!). A line to say "Run the Protoc generator on the Extension proto" should suffice.

Comment (noel-yap) [reply jprotoc/README.md]: Done.

**Pull request merged**